go/netutil/conn.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2018 The Vitess Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package netutil
+
+import (
+	"net"
+	"time"
+)
+
+var _ net.Conn = (*ConnWithTimeout)(nil)
+
+type ConnWithTimeout struct {
+	conn         net.Conn
+	readTimeout  time.Duration
+	writeTimeout time.Duration
+}
+
+// NewConnWithTimeout wraps a net.Conn with read and write deadilnes.
+func NewConnWithTimeout(conn net.Conn, readTimeout time.Duration, writeTimeout time.Duration) ConnWithTimeout {
+	return ConnWithTimeout{conn: conn, readTimeout: readTimeout, writeTimeout: writeTimeout}
+}
+
+// Implementation of the Conn interface.
+
+// Read sets a read deadilne and delegates to conn.Read.
+func (c ConnWithTimeout) Read(b []byte) (int, error) {
+	if err := c.conn.SetReadDeadline(time.Now().Add(c.readTimeout)); err != nil {
+		return 0, err
+	}
+	return c.conn.Read(b)
+}
+
+// Write sets a write deadline and delagates to conn.Write
+func (c ConnWithTimeout) Write(b []byte) (int, error) {
+	if err := c.conn.SetWriteDeadline(time.Now().Add(c.writeTimeout)); err != nil {
+		return 0, err
+	}
+	return c.conn.Write(b)
+}
+
+// Cloase is a passthrough to net.Conn Close
+func (c ConnWithTimeout) Close() error {
+	return c.conn.Close()
+}
+
+// LocalAddr implements the Conn LocalAddr method.
+func (c ConnWithTimeout) LocalAddr() net.Addr {
+	return c.conn.LocalAddr()
+}
+
+// RemoteAddr is a passthrough to net.Conn RemoteAddr
+func (c ConnWithTimeout) RemoteAddr() net.Addr {
+	return c.conn.RemoteAddr()
+}
+
+// SetDeadline implements the Conn SetDeadline method.
+func (c ConnWithTimeout) SetDeadline(t time.Time) error {
+	panic("can't call SetDeadline for ConnWithTimeout")
```

Comment:
@demmer - after a second thought, I wonder if we shouldn't enforce this. If someone want's to use a different timeout for a particular call, don't see why it shouldn't be allowed.